File: python/src/codex_client/transport.py
from __future__ import annotations

import asyncio
import os
from collections import deque
from collections.abc import AsyncIterator, Mapping, Sequence
from typing import Protocol, runtime_checkable

from .errors import ProtocolStreamError, TransportClosedError

_QUEUE_EOF = object()


@runtime_checkable
class Transport(Protocol):
    async def send(self, data: str) -> None: ...

    def __aiter__(self) -> AsyncIterator[str]: ...

    async def close(self) -> None: ...


class StdioTransport:
    def __init__(
        self,
        command: str = "codex",
        args: Sequence[str] = ("app-server",),
        cwd: str | None = None,
        env: Mapping[str, str] | None = None,
    ) -> None:
        self._command = command
        self._args = tuple(args)
        self._cwd = cwd
        self._env = dict(env) if env is not None else None
        self._stderr_tail: deque[str] = deque(maxlen=50)
        self._queue: asyncio.Queue[object] = asyncio.Queue()
        self._process: asyncio.subprocess.Process | None = None
        self._stdout_task: asyncio.Task[None] | None = None
        self._stderr_task: asyncio.Task[None] | None = None
        self._start_lock = asyncio.Lock()
        self._close_lock = asyncio.Lock()
        self._closed = False

    async def __aenter__(self) -> "StdioTransport":
        await self._ensure_started()
        return self

    async def __aexit__(self, exc_type, exc, tb) -> None:
        await self.close()

    async def send(self, data: str) -> None:
        if "\n" in data or "\r" in data:
            raise ValueError("StdioTransport.send expects a single line without newlines")

        await self._ensure_started()

        if self._closed:
            raise TransportClosedError("Transport is closed")

        process = self._require_process()
        if process.stdin is None:
            raise TransportClosedError("Transport stdin is not available")

        try:
            process.stdin.write(f"{data}\n".encode("utf-8"))
            await process.stdin.drain()
        except (BrokenPipeError, ConnectionResetError) as exc:
            raise TransportClosedError("Transport stdin is closed") from exc

    def __aiter__(self) -> AsyncIterator[str]:
        return self._iterate()

    async def close(self) -> None:
        async with self._close_lock:
            if self._closed:
                return

            self._closed = True
            process = self._process

            if process is not None and process.stdin is not None:
                process.stdin.close()
                try:
                    await process.stdin.wait_closed()
                except (BrokenPipeError, ConnectionResetError):
                    pass

            if process is not None and process.returncode is None:
                process.terminate()
                try:
                    await asyncio.wait_for(process.wait(), timeout=2.0)
                except asyncio.TimeoutError:
                    process.kill()
                    await process.wait()

            if self._stdout_task is not None:
                await self._stdout_task
            if self._stderr_task is not None:
                await self._stderr_task

    async def _ensure_started(self) -> None:
        if self._process is not None:
            return

        async with self._start_lock:
            if self._process is not None:
                return
            if self._closed:
                raise TransportClosedError("Transport is closed")

            env = None
            if self._env is not None:
                env = os.environ.copy()
                env.update(self._env)

            self._process = await asyncio.create_subprocess_exec(
                self._command,
                *self._args,
                cwd=self._cwd,
                env=env,
                stdin=asyncio.subprocess.PIPE,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
            )
            self._stdout_task = asyncio.create_task(self._pump_stdout())
            self._stderr_task = asyncio.create_task(self._pump_stderr())

    async def _iterate(self) -> AsyncIterator[str]:
        await self._ensure_started()

        while True:
            item = await self._queue.get()
            if item is _QUEUE_EOF:
                return
            if isinstance(item, BaseException):
                raise item
            yield str(item)

    async def _pump_stdout(self) -> None:
        process = self._require_process()
        assert process.stdout is not None

        while True:
            line = await process.stdout.readline()
            if not line:
                break
            await self._queue.put(line.decode("utf-8").rstrip("\r\n"))

        returncode = await process.wait()
        if self._closed:
            await self._queue.put(_QUEUE_EOF)
            return

        await self._queue.put(
            ProtocolStreamError(self._format_process_exit_message(returncode))
        )

    async def _pump_stderr(self) -> None:
        process = self._require_process()
        assert process.stderr is not None

        while True:
            line = await process.stderr.readline()
            if not line:
                return
            self._stderr_tail.append(line.decode("utf-8", errors="replace").rstrip())

    def _format_process_exit_message(self, returncode: int) -> str:
        message = f"Transport process exited unexpectedly with code {returncode}"
        if self._stderr_tail:
            message = f"{message}. stderr tail:\n" + "\n".join(self._stderr_tail)
        return message

    def _require_process(self) -> asyncio.subprocess.Process:
        if self._process is None:
            raise TransportClosedError("Transport process has not been started")
        return self._process

File: typescript/src/transport.ts
import { spawn, type ChildProcessWithoutNullStreams } from "node:child_process";
import { createInterface } from "node:readline";

import { ProtocolStreamError, TransportClosedError } from "./errors.js";

export interface Transport extends AsyncIterable<string> {
  send(data: string): Promise<void>;
  close(): Promise<void>;
}

export interface StdioTransportOptions {
  command?: string;
  args?: readonly string[];
  cwd?: string;
  env?: Readonly<Record<string, string>>;
  killTimeoutMs?: number;
}

type QueueItem =
  | { type: "frame"; frame: string }
  | { type: "error"; error: Error }
  | { type: "eof" };

const DEFAULT_KILL_TIMEOUT_MS = 2_000;

export class StdioTransport implements Transport {
  private readonly command: string;
  private readonly args: readonly string[];
  private readonly cwd: string | undefined;
  private readonly env: Readonly<Record<string, string>> | undefined;
  private readonly killTimeoutMs: number;
  private readonly stderrTail: string[] = [];
  private readonly queue = new AsyncQueue<QueueItem>();
  private process: ChildProcessWithoutNullStreams | null = null;
  private startupPromise: Promise<void> | null = null;
  private stdoutTask: Promise<void> | null = null;
  private stderrTask: Promise<void> | null = null;
  private closePromise: Promise<void> | null = null;
  private closed = false;

  constructor(options: StdioTransportOptions = {}) {
    this.command = options.command ?? "codex";
    this.args = options.args ?? ["app-server"];
    this.cwd = options.cwd;
    this.env = options.env;
    this.killTimeoutMs = options.killTimeoutMs ?? DEFAULT_KILL_TIMEOUT_MS;
  }

  async send(data: string): Promise<void> {
    if (data.includes("\n") || data.includes("\r")) {
      throw new ValueError("StdioTransport.send expects a single line without newlines");
    }

    await this.ensureStarted();

    if (this.closed) {
      throw new TransportClosedError("Transport is closed");
    }

    const child = this.requireProcess();
    const stdin = child.stdin;
    if (stdin.destroyed) {
      throw new TransportClosedError("Transport stdin is closed");
    }

    await new Promise<void>((resolve, reject) => {
      let settled = false;

      const finishResolve = () => {
        if (settled) {
          return;
        }

        settled = true;
        cleanup();
        resolve();
      };

      const finishReject = (error: unknown) => {
        if (settled) {
          return;
        }

        settled = true;
        cleanup();
        reject(new TransportClosedError("Transport stdin is closed", { cause: asError(error) }));
      };

      const onError = (error: Error) => {
        finishReject(error);
      };

      const onDrain = () => {
        finishResolve();
      };

      const cleanup = () => {
        stdin.off("error", onError);
        stdin.off("drain", onDrain);
      };

      stdin.once("error", onError);
      const canWrite = stdin.write(`${data}\n`, "utf8", (error) => {
        if (error) {
          finishReject(error);
          return;
        }

        if (canWrite) {
          finishResolve();
        }
      });

      if (!canWrite) {
        stdin.once("drain", onDrain);
      }
    });
  }

  async close(): Promise<void> {
    if (this.closePromise !== null) {
      return this.closePromise;
    }

    this.closed = true;
    this.closePromise = this.closeInternal();
    return this.closePromise;
  }

  async *[Symbol.asyncIterator](): AsyncIterator<string> {
    await this.ensureStarted();

    while (true) {
      const item = await this.queue.dequeue();
      if (item.type === "eof") {
        return;
      }
      if (item.type === "error") {
        throw item.error;
      }
      yield item.frame;
    }
  }

  private async closeInternal(): Promise<void> {
    const child = this.process;
    if (child === null) {
      return;
    }

    const exitPromise = this.waitForExit(child);
    if (!child.stdin.destroyed) {
      child.stdin.end();
      await onceStreamFinished(child.stdin);
    }

    const exitedAfterStdinClose = await promiseWithTimeout(exitPromise, this.killTimeoutMs);
    if (exitedAfterStdinClose === null && child.exitCode === null && child.signalCode === null) {
      child.kill("SIGTERM");
      const exitedAfterTerminate = await promiseWithTimeout(exitPromise, this.killTimeoutMs);
      if (exitedAfterTerminate === null && child.exitCode === null && child.signalCode === null) {
        child.kill("SIGKILL");
        await exitPromise;
      }
    }

    if (this.stdoutTask !== null) {
      await this.stdoutTask;
    }
    if (this.stderrTask !== null) {
      await this.stderrTask;
    }
  }

  private async ensureStarted(): Promise<void> {
    if (this.process !== null) {
      return;
    }

    if (this.startupPromise !== null) {
      return this.startupPromise;
    }

    if (this.closed) {
      throw new TransportClosedError("Transport is closed");
    }

    this.startupPromise = this.startProcess();
    try {
      await this.startupPromise;
    } finally {
      this.startupPromise = null;
    }
  }

  private async startProcess(): Promise<void> {
    const child = spawn(this.command, this.args, {
      cwd: this.cwd,
      env: this.env === undefined ? process.env : { ...process.env, ...this.env },
      stdio: ["pipe", "pipe", "pipe"],
    });

    this.process = child;

    try {
      await new Promise<void>((resolve, reject) => {
        const onSpawn = () => {
          cleanup();
          resolve();
        };

        const onError = (error: Error) => {
          cleanup();
          reject(error);
        };

        const cleanup = () => {
          child.off("spawn", onSpawn);
          child.off("error", onError);
        };

        child.once("spawn", onSpawn);
        child.once("error", onError);
      });
    } catch (error: unknown) {
      this.process = null;
      throw error;
    }

    child.on("error", (error) => {
      if (!this.closed) {
        this.queue.enqueue({ type: "error", error });
      }
    });

    this.stdoutTask = this.pumpStdout(child);
    this.stderrTask = this.pumpStderr(child);
  }

  private async pumpStdout(child: ChildProcessWithoutNullStreams): Promise<void> {
    const reader = createInterface({
      input: child.stdout,
      crlfDelay: Infinity,
    });

    try {
      for await (const line of reader) {
        this.queue.enqueue({ type: "frame", frame: line });
      }

      const { code, signal } = await this.waitForExit(child);
      if (this.closed) {
        this.queue.enqueue({ type: "eof" });
        return;
      }

      this.queue.enqueue({
        type: "error",
        error: new ProtocolStreamError(this.formatProcessExitMessage(code, signal)),
      });
    } catch (error: unknown) {
      if (this.closed) {
        this.queue.enqueue({ type: "eof" });
        return;
      }

      this.queue.enqueue({ type: "error", error: asError(error) });
    } finally {
      reader.close();
    }
  }

  private async pumpStderr(child: ChildProcessWithoutNullStreams): Promise<void> {
    const reader = createInterface({
      input: child.stderr,
      crlfDelay: Infinity,
    });

    try {
      for await (const line of reader) {
        this.stderrTail.push(line);
        if (this.stderrTail.length > 50) {
          this.stderrTail.shift();
        }
      }
    } finally {
      reader.close();
    }
  }

  private formatProcessExitMessage(
    code: number | null,
    signal: NodeJS.Signals | null,
  ): string {
    const exitDetail =
      code === null
        ? `signal ${signal ?? "unknown"}`
        : `code ${code}`;

    let message = `Transport process exited unexpectedly with ${exitDetail}`;
    if (this.stderrTail.length > 0) {
      message += `. stderr tail:\n${this.stderrTail.join("\n")}`;
    }
    return message;
  }

  private requireProcess(): ChildProcessWithoutNullStreams {
    if (this.process === null) {
      throw new TransportClosedError("Transport process has not been started");
    }
    return this.process;
  }

  private waitForExit(
    child: ChildProcessWithoutNullStreams,
  ): Promise<{ code: number | null; signal: NodeJS.Signals | null }> {
    if (child.exitCode !== null || child.signalCode !== null) {
      return Promise.resolve({
        code: child.exitCode,
        signal: child.signalCode,
      });
    }

    return new Promise((resolve) => {
      child.once("exit", (code, signal) => {
        resolve({ code, signal });
      });
    });
  }
}

class AsyncQueue<T> {
  private readonly items: T[] = [];
  private readonly waiters: Array<(value: T) => void> = [];

  enqueue(value: T): void {
    const waiter = this.waiters.shift();
    if (waiter !== undefined) {
      waiter(value);
      return;
    }

    this.items.push(value);
  }

  dequeue(): Promise<T> {
    const value = this.items.shift();
    if (value !== undefined) {
      return Promise.resolve(value);
    }

    return new Promise((resolve) => {
      this.waiters.push(resolve);
    });
  }
}

class ValueError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ValueError";
  }
}

function asError(error: unknown): Error {
  return error instanceof Error ? error : new Error(String(error));
}

function onceStreamFinished(
  stream: NodeJS.WritableStream & { writableFinished?: boolean; destroyed?: boolean },
): Promise<void> {
  if (stream.writableFinished || stream.destroyed) {
    return Promise.resolve();
  }

  return new Promise((resolve) => {
    stream.once("finish", () => resolve());
    stream.once("close", () => resolve());
    stream.once("error", () => resolve());
  });
}

async function promiseWithTimeout<T>(promise: Promise<T>, timeoutMs: number): Promise<T | null> {
  let timer: NodeJS.Timeout | undefined;

  try {
    return await Promise.race([
      promise,
      new Promise<null>((resolve) => {
        timer = setTimeout(() => resolve(null), timeoutMs);
      }),
    ]);
  } finally {
    if (timer !== undefined) {
      clearTimeout(timer);
    }
  }
}
